File: haskell/bob/src/Bob.hs
module Bob (responseFor) where
import Data.Char (isSpace, isLetter, isUpper)

-- Bob exercise

responseFor :: String -> String
responseFor input
  | null text = "Fine. Be that way!"
  | isShouting && isAsking = "Calm down, I know what I'm doing!"
  | isShouting = "Whoa, chill out!"
  | isAsking = "Sure."
  | otherwise = "Whatever."
  where
    text = filter (not . isSpace) input
    letters = filter isLetter text
    isShouting = all isUpper letters && any isUpper letters
    isAsking = last text == '?'

File: cpp/bob/bob.cpp
#include <regex>
#include "bob.h"

namespace bob
{
  string hey(const string &s)
  {
    bitset<3> flags;
    flags[0] = regex_match(s, regex(".*\\?\\s*$"));
    flags[1] = regex_match(s, regex(".*[A-Z]+.*")) && !regex_match(s, regex(".*[a-z]+.*"));
    flags[2] = regex_match(s, regex("^\\s*$"));

    switch (flags.to_ulong())
    {
    case 0b001:
      return "Sure.";
    case 0b010:
      return "Whoa, chill out!";
    case 0b011:
      return "Calm down, I know what I'm doing!";
    case 0b100:
      return "Fine. Be that way!";
    default:
      return "Whatever.";
      break;
    }
  }
}
